pkg/operator/configobservation/auth/external_oidc.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/configobserver/featuregates"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resourcesynccontroller"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apiserver/pkg/apis/apiserver"
+	"k8s.io/klog/v2"
+)
+
+const (
+	apiServerArgumentsPath = "apiServerArguments"
+	argAuthConfig          = "authentication-config"
+
+	SourceAuthConfigCMNamespace = "openshift-config-managed"
+	AuthConfigCMName            = "auth-config"
+	staticAuthConfigPath        = "/etc/kubernetes/static-pod-resources/configmaps/" + AuthConfigCMName + "/auth-config.json"
+)
+
+func NewObserveExternalOIDC(featureGateAccessor featuregates.FeatureGateAccess) configobserver.ObserveConfigFunc {
+	return (&externalOIDC{
+		featureGateAccessor: featureGateAccessor,
+	}).ObserveExternalOIDC
+}
+
+type externalOIDC struct {
+	featureGateAccessor featuregates.FeatureGateAccess
+}
+
+// ObserveExternalOIDC observes the authentication.config/cluster resource
+// and if the type field is set to OIDC, it configures an external OIDC provider
+// to the KAS pods by setting the --authentication-config apiserver argument. It also
+// takes care of synchronizing the structured auth config file into the apiserver's namespace
+// so that it gets mounted as a static file on each node.
+func (o *externalOIDC) ObserveExternalOIDC(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	if o.featureGateAccessor == nil || !o.featureGateAccessor.AreInitialFeatureGatesObserved() {
+		// if we haven't observed featuregates yet, return the existing
+		return existingConfig, nil
+	}
+
+	featureGates, err := o.featureGateAccessor.CurrentFeatureGates()
+	if err != nil {
+		return existingConfig, []error{err}
+	}
+
+	if !featureGates.Enabled(features.FeatureGateExternalOIDC) {
+		return existingConfig, nil
+	}
+
+	listers := genericListers.(configobservation.Listers)
+	auth, err := listers.AuthConfigLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		klog.Warningf("authentications.config.openshift.io/cluster: not found")
+		return existingConfig, nil
+	} else if err != nil {
+		return existingConfig, []error{err}
+	}
+
+	if auth.Spec.Type != configv1.AuthenticationTypeOIDC {
+		if _, err := listers.ConfigMapLister().ConfigMaps(operatorclient.TargetNamespace).Get(AuthConfigCMName); errors.IsNotFound(err) {
+			return nil, nil
+
+		} else if err != nil {
+			return existingConfig, []error{fmt.Errorf("failed to get configmap %s/%s: %v", operatorclient.TargetNamespace, AuthConfigCMName, err)}
+		}
```

Comment:
Will the following call to `syncConfigMap` do effectively the same thing? What's the advantage of short-circuiting here?

Reply:
I wanted to avoid another sync when we know we don't have that CM anymore, mainly because the syncer seems to be [doing a Get API call](https://github.com/liouk/cluster-kube-apiserver-operator/blob/5cdfb5e78892398989fcc6bc3efa7f09030e081c/vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/core.go#L472).